src/attr_color.rs

```diff
@@ -0,0 +1,76 @@
+use glib::translate::from_glib_none;
+use glib::translate::from_glib_full;
+use glib::translate::{FromGlibPtrFull, FromGlibPtrNone, Stash, ToGlibPtr};
+use pango_sys;
+use Attribute;
+use Color;
+
+#[doc(hidden)]
+impl<'a> ToGlibPtr<'a, *mut pango_sys::PangoAttrColor> for &'a AttrColor {
+    type Storage = &'a AttrColor;
+
+    fn to_glib_none(&self) -> Stash<'a, *mut pango_sys::PangoAttrColor, Self> {
+        Stash(self.0, *self)
+    }
+}
+
+#[doc(hidden)]
+impl FromGlibPtrNone<*mut pango_sys::PangoAttrColor> for AttrColor {
+    unsafe fn from_glib_none(ptr: *mut pango_sys::PangoAttrColor) -> Self {
+        assert!(!ptr.is_null());
+        AttrColor(ptr)
+    }
+}
+
+#[doc(hidden)]
+impl FromGlibPtrFull<*mut pango_sys::PangoAttrColor> for AttrColor {
+    unsafe fn from_glib_full(ptr: *mut pango_sys::PangoAttrColor) -> Self {
+        assert!(!ptr.is_null());
+        AttrColor(ptr)
+    }
+}
+
+#[doc(hidden)]
+impl FromGlibPtrNone<*const pango_sys::PangoAttrColor> for AttrColor {
+    unsafe fn from_glib_none(ptr: *const pango_sys::PangoAttrColor) -> Self {
+        assert!(!ptr.is_null());
+        AttrColor(ptr as *mut _)
+    }
+}
+
+#[doc(hidden)]
+impl FromGlibPtrFull<*const pango_sys::PangoAttrColor> for AttrColor {
+    unsafe fn from_glib_full(ptr: *const pango_sys::PangoAttrColor) -> Self {
+        assert!(!ptr.is_null());
+        AttrColor(ptr as *mut _)
+    }
+}
+
+#[derive(Debug)]
+pub struct AttrColor(*mut pango_sys::PangoAttrColor);
+
+impl AttrColor {
+    pub fn get_color(&self) -> Color {
+        unsafe {
+            let color: *const pango_sys::PangoColor = &(*self.0).color;
+            from_glib_none(color)
+        }
+    }
+}
+
+impl From<Attribute> for AttrColor {
+    fn from(attribute: Attribute) -> Self {
+        unsafe {
+            let attr_color = std::mem::transmute::<*const pango_sys::PangoAttribute, *const pango_sys::PangoAttrColor>(attribute.to_glib_none().0);
+            from_glib_full(attr_color)
+        }
+    }
+}
+
+impl PartialEq for AttrColor {
+    fn eq(&self, other: &AttrColor) -> bool {
+        self.0 == other.0
```

Comment:
This should probably call `pango_attr_equal()`

And a `Drop` impl that calls `pango_attr_destroy()` should probably exist too, and a `Clone` impl that calls `pango_attr_copy()`

I think the attrs should probably be implemented like the patterns in `cairo`. The generic `pango_attr` API works on all of them and then there are "subclasses" of the generic `PangoAttr` type.

Reply:
@sdroege would you be able to provide me with some links to the code that I could use as references? Especially to the patterns in `cairo`. Thanks.

Reply:
Take a look here: https://github.com/gtk-rs/cairo/blob/master/src/patterns.rs

There's a base `Pattern` type and specific `struct FooPattern(Pattern)` that provide specific API. `Pattern` impls `Drop` for memory management, and there are `try_from()` and `deref()` impls to get from one to another.

Reply:
This still applies to the PR and requires some more work

Reply:
@GuillaumeGomez See above, that's the main missing part here. Let's just defer that to a future release, it involves quite some work and figuring out a nice API. I wouldn't want to merge that just before a release but rather give people some time to try it first.

Reply:
I removed all sub-Attribute types. The API is shitty as hell.

Reply:
Ok that also makes sense, yes. What a weird API :) So you can create all kinds of attributes but you can't look at their values afterwards, only at the type.

Reply:
Hmm no. Why do you think they are not supposed to be used? They're public API and even documented, so access to the fields of the different types seems to be expected. And you actually *need* that when iterating over attributes, otherwise they're completely useless.

Reply:
This really has to be handled exactly the same way as the patterns in cairo. No shortcut here, sorry :)

Reply:
No, definitely not. The API docs don't show any way to get a specific attribute, even when creating one from a specific method (like `new_foreground` or whatever). At least, providing these functions allow people to create attributes even if they have to remove them in order to update them.

Reply:
No, definitely not. The API docs don't show any way to get a specific attribute, even when creating one from a specific method (like `new_foreground` or whatever). At least, providing these functions allow people to create attributes even if they have to remove them in order to update them.

Reply:
It's not about updating (I'm fine with having the immutable), but you need access to the fields for actually making use of this API at all. How would you otherwise read the color of the background attribute for example, the attribute you might've gotten from an attribute iterator or attribute list in a generic way.

You need a way to cast from a generic attribute to the specific type based on the type of it (which you can get from the `klass` field). You can simply cast a `PangoAttribute*` to an `PangoAttrColor*` if it's of the corresponding type, and that's how this API works.

Reply:
Yes and it's horrible. But I guess I'll implement a widget upgrade/downgrade equivalent...

Reply:
Yeah unfortunately. That's why I said that this should probably wait for after this release. It involves writing quite some code and API.

The way how it is right now here is not useful for actually using the attributes.

Reply:
But it at least allows to use them. Currently it's not possible.

Reply:
As usable as if we had bindings for `GObject` but not `GtkButton` :)

Reply:
More like a `Widget`and not `Button`. ;) From my point of view, we should merge this and implement what's left in the next release.

Reply:
It's useless as it is and will only cause bug reports from people trying to figure out how to use it and not understanding :) And worst case we introduce a new actual bug here that slipped through the review.

I would rather wait until after the release, there's no advantage in having this in this release already and only the risk of introducing a new bug.

Reply:
You mean except being able to create the attributes which we cannot currently? :p

It seems it's more of a "create and never modify" API from what I saw. So I guess it's mostly fine. Also, where could they get bugs? We don't access to anything...

Reply:
What do you think about this @EPashkin ?

Reply:
Waiting for his opinion then :) Maybe I'm just a bit too strict here.